### PR TITLE
Remove/defer torchvision imports

### LINKED ISSUE #936: Defer `torchvision` import?
In the quest to minimise `deepinv` overhead as much as possible, we still require the eager import of `torchvision` in deepinv as it's used in many places.

However, torchvision is still quite a heavy import:

Image

It would be good to review how much we use torchvision/assume the user is already using torchvision. If a lot, then no point in removing the eager import. If less so, then we could try to remove it.

### PR DESCRIPTION
**Pull request by Andrewwango**
Closes #936 . See that issue for motivation

This removes all top-level eager torchvision imports from the main library. Now, import deepinv does not import torchvision eagerly anymore.

All torchvision imports in examples and tests are unchanged.

This PR also breaks some long import chains by forcing lazy imports of some external libs like PIL, h5py as well as lazy imports of physics inside models etc.

NOTE for future reference: removing torchvision altogether is also an option, in order to reduce the install dependencies, and remove some layers of abstraction. However, this is out of scope.

TLDR: import time cut in half, with minimal overhead on top of torch!

image

cc @nucli-vicky 

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

**Comment by Andrewwango** [comment]
@nucli-vicky @jscanvic would appreciate a review please! I went a bit obsessive over reducing import time.

File: pyproject.toml [hunk 1]
```diff
@@ -27,8 +27,8 @@ dependencies = [
     "matplotlib",
     "tqdm",
     "torch>=2.2.0",
-    "torchvision",
     "torchmetrics",
+    "torchvision",
```
Comment (jscanvic) [inline pyproject.toml:31]: Why?

File: deepinv/__init__.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 from .__about__ import *
 import torch
-import torchvision as _torchvision  # early loading
```
Comment (jscanvic) [inline deepinv/__init__.py:3]: Nice

File: deepinv/utils/io.py [hunk 3]
```diff
@@ -347,6 +351,8 @@ def _process_array(arr: np.ndarray) -> torch.Tensor:
         )  # Remove single-band dimension if exists
         if arr.ndim == 2:
             if arr.is_complex():
+                from deepinv.utils.mixins import MRIMixin
```
Comment (jscanvic) [inline deepinv/utils/io.py:354]: Why?

File: deepinv/utils/io.py [hunk 4]
```diff
@@ -71,21 +69,24 @@ def get_cache_home() -> Path:
     return path
 
 
-class DownloadError(requests.exceptions.RequestException):
-    r"""Raised when a network download initiated by deepinv fails.
+class DownloadError(IOError):
```
Comment (jscanvic) [inline deepinv/utils/io.py:72]:
```suggestion
class DownloadError(OSError):
```

OSError is the canonical alias of IOError

https://docs.python.org/3/library/exceptions.html#IOError

Comment (jscanvic) [inline deepinv/utils/io.py:72]: Are we sure this is a minor/negligible breakage to change the inheritance chain? I guess it is but I'm just making sure

File: deepinv/utils/__init__.py [hunk 5]
```diff
@@ -7,7 +7,6 @@
     plot_inset,
     plot_videos,
     save_videos,
-    make_grid,
```
Comment (jscanvic) [inline deepinv/utils/__init__.py:10]: Put it as breaking in the changelog?

File: deepinv/transform/rotate.py [hunk 6]
```diff
@@ -57,6 +61,10 @@ def __init__(
             else interpolation_mode
         )
 
+        from torchvision.transforms.functional import rotate as torchvision_rotate
+
+        self.torchvision_rotate = torchvision_rotate  # lazy import
```
Comment (jscanvic) [inline deepinv/transform/rotate.py:66]:
```suggestion
        self._torchvision_rotate = torchvision_rotate  # lazy import
```

File: deepinv/transform/rotate.py [hunk 7]
```diff
@@ -86,7 +94,7 @@ def _transform(
         """
         return torch.cat(
             [
-                rotate(
+                self.torchvision_rotate(
```
Comment (jscanvic) [inline deepinv/transform/rotate.py:97]:
```suggestion
                self._torchvision_rotate(
```

File: deepinv/optim/data_fidelity.py [hunk 8]
```diff
@@ -350,6 +349,10 @@ def __init__(self, sigma: float = 1.0, threshold: float = 1.0):
         self.norm = 1 / (sigma**2)
         self.modulo_round = lambda x: x - threshold * torch.round(x / threshold)
 
+        from deepinv.physics.functional import dct_2d, idct_2d
+
+        self.dct_2d, self.idct_2d = dct_2d, idct_2d  # lazy import
```
Comment (jscanvic) [inline deepinv/optim/data_fidelity.py:354]:
```suggestion
        self._dct_2d, self._idct_2d = dct_2d, idct_2d  # lazy import
```

File: deepinv/optim/data_fidelity.py [hunk 9]
```diff
@@ -539,14 +542,14 @@ def prox(
                 - (torch.cos(torch.pi * I / MX) + torch.cos(torch.pi * J / NX))
             )
 
-        dct_psi = dct_2d(psi, norm="ortho")
+        dct_psi = self.dct_2d(psi, norm="ortho")
```
Comment (jscanvic) [inline deepinv/optim/data_fidelity.py:545]:
```suggestion
        dct_psi = self._dct_2d(psi, norm="ortho")
```

File: deepinv/optim/data_fidelity.py [hunk 10]
```diff
@@ -539,14 +542,14 @@ def prox(
...
 
         denom = denom.to(psi.device)
         denom[..., 0, 0] = 1  # avoid division by zero
 
         dct_phi = dct_psi / denom
 
-        phi = idct_2d(dct_phi, norm="ortho")
+        phi = self.idct_2d(dct_phi, norm="ortho")
```
Comment (jscanvic) [inline deepinv/optim/data_fidelity.py:552]:
```suggestion
        phi = self._idct_2d(dct_phi, norm="ortho")
```

File: deepinv/datasets/kohler.py [hunk 11]
```diff
@@ -154,6 +155,8 @@ def download(cls, root: str | Path = None, remove_finished: bool = False) -> Non
                 from deepinv.datasets import Kohler
                 Kohler.download("datasets/Kohler")
         """
+        from torchvision.datasets.utils import download_and_extract_archive
```
Comment (nucli-vicky) [inline deepinv/datasets/kohler.py:158]: We can probably have the same function with existing `deepinv` functions in `datasets/utils.py` quite easily: https://github.com/pytorch/vision/blob/da446275eccae6cd00923080d9cc91f1f3a73744/torchvision/datasets/utils.py#L374-L391

File: deepinv/loss/sup.py [hunk 12]
```diff
@@ -6,7 +6,6 @@
 from deepinv.loss.loss import Loss
 from deepinv.loss.metric.metric import Metric
 from deepinv.models.base import Reconstructor
-from deepinv.physics.forward import Physics
```
Comment (nucli-vicky) [inline deepinv/loss/sup.py:9]: good catch :)

File: deepinv/models/bm3d.py [hunk 13]
```diff
@@ -168,6 +167,8 @@ def _get_wavelet_matrix_torch(
     def _get_dct_matrix(
         self, n: int, norm: str = "ortho", device: torch.device | None = None
     ) -> tuple[Tensor, Tensor]:
+        from deepinv.physics.functional import dct  # lazy import at init
```
Comment (nucli-vicky) [inline deepinv/models/bm3d.py:170]: I'm a bit confused: in some classes, you added that we import during the `__init__` of the class, and in others (like here), import within the class method that needs the import.

Is there a consistent logic for this right now?

File: deepinv/models/equivariant.py [hunk 14]
```diff
@@ -141,6 +140,12 @@ def __init__(
         self.transform = transform
         self.eval_transform = eval_transform
 
+        from deepinv.physics.virtual import VirtualLinearPhysics
+
+        self.virtual_linear_physics = (
+            VirtualLinearPhysics  # lazy import to break import chains
+        )
+
```
Comment (nucli-vicky) [inline deepinv/models/equivariant.py:148]: For example, this import can also be done at `forward`. I don't mind either (although importing within the method that needs it is less code), just wondering what the logic is right now :)

File: deepinv/utils/__init__.py (hunk 5 above)
Comment (nucli-vicky) [inline deepinv/utils/__init__.py:10]: this would break user code doing 'from deepinv.utils import make_grid`.

I think it makes sense to remove, but we should either (i) clearly note this in changelog or (ii) write `make_grid` within `deepinv`

**Review by nucli-vicky — COMMENTED** [review]
Thanks @Andrewwango 

I especially like the `TYPE_CHECKING` gated imports when a class isn't needed besides type checking :)

I know there is some disagreement in styleguides w.r.t the use of lazy imports, but personally I'm fine with them here as it is for performance reasons.

Some additional comments, aside from the bugs causing current test errors

File: deepinv/models/bm3d.py (hunk 13 above)
Comment (Andrewwango) [reply deepinv/models/bm3d.py:170]: good point. I was not consistent, I was just doing whatever felt most elegant per class, which is purely subjective. I think the logic should be: external imports (torchvision, PIL) should be imported at least once a init, and keep a reference to the object (otherwise ruff will get angry we don't use the import) - or, even don't keep a reference, and suppress ruff. Internal deepinv imports, we can do it in the methods

File: deepinv/models/equivariant.py (hunk 14 above)
Comment (Andrewwango) [reply deepinv/models/equivariant.py:148]: agreed, can do at forward

**Review by jscanvic — CHANGES_REQUESTED** [review]
Thanks Andrew!

A bit of a messy review (sorry!) I did back then and wasn't 100% happy with - feel free to ignore some of the comments

Once you've been through it & CI passes happy to approve
